contracts/src/libraries/codec/BatchHeaderV1Codec.sol

```diff
@@ -0,0 +1,230 @@
+        assembly {
+            mstore(add(batchPtr, 9), shl(192, _l1MessagePopped))
+        }
+    }
+
+    /// @notice Store the total number of L1 messages popped after current batch to batch header.
+    /// @dev Because this function can overwrite the subsequent fields, it must be called before
+    /// `storeDataHash`.
+    /// @param batchPtr The start memory offset of the batch header in memory.
+    /// @param _totalL1MessagePopped The total number of L1 messages popped after current batch.
+    function storeTotalL1MessagePopped(uint256 batchPtr, uint256 _totalL1MessagePopped) internal pure {
+        assembly {
+            mstore(add(batchPtr, 17), shl(192, _totalL1MessagePopped))
+        }
+    }
+
+    /// @notice Store the data hash of batch header.
+    /// @param batchPtr The start memory offset of the batch header in memory.
+    /// @param _dataHash The data hash.
+    function storeDataHash(uint256 batchPtr, bytes32 _dataHash) internal pure {
+        assembly {
+            mstore(add(batchPtr, 25), _dataHash)
+        }
+    }
+
+    /// @notice Store the parent batch hash of batch header.
+    /// @param batchPtr The start memory offset of the batch header in memory.
+    /// @param _blobVersionedHash The versioned hash of the blob with this batch’s data.
+    function storeBlobVersionedHash(uint256 batchPtr, bytes32 _blobVersionedHash) internal pure {
+        assembly {
+            mstore(add(batchPtr, 57), _blobVersionedHash)
+        }
+    }
+
+    /// @notice Store the parent batch hash of batch header.
+    /// @param batchPtr The start memory offset of the batch header in memory.
+    /// @param _parentBatchHash The parent batch hash.
+    function storeParentBatchHash(uint256 batchPtr, bytes32 _parentBatchHash) internal pure {
+        assembly {
+            mstore(add(batchPtr, 89), _parentBatchHash)
+        }
+    }
+
+    /// @notice Store the skipped L1 message bitmap of batch header.
+    /// @param batchPtr The start memory offset of the batch header in memory.
+    /// @param _skippedL1MessageBitmap The skipped L1 message bitmap.
+    function storeSkippedBitmap(uint256 batchPtr, bytes calldata _skippedL1MessageBitmap) internal pure {
+        assembly {
+            calldatacopy(
+                add(batchPtr, BATCH_HEADER_FIXED_LENGTH),
+                _skippedL1MessageBitmap.offset,
+                _skippedL1MessageBitmap.length
+            )
+        }
+    }
+
+    /// @notice Compute the batch hash.
+    /// @dev Caller should make sure that the encoded batch header is correct.
+    ///
+    /// @param batchPtr The start memory offset of the batch header in memory.
+    /// @param length The length of the batch.
+    /// @return _batchHash The hash of the corresponding batch.
+    function computeBatchHash(uint256 batchPtr, uint256 length) internal pure returns (bytes32 _batchHash) {
+        // in the current version, the hash is: keccak(BatchHeader without timestamp)
```

Comment:
What's the case with timestamps?

Reply:
we've discussed about whether to add `block.timestamp` to the batch header hash. So in current version (v0 and v1), we won't.